### PR TITLE
sl-52: sites page

### PR DESCRIPTION
**Pull request by Alyona-Pisotska**
photo_2023-03-23 11 56 29
photo_2023-03-23 11 56 37

### CODE REVIEW

**Comment by what1s1ove** [comment]
merge `development` into your branch

File: frontend/src/pages/sites/components/site-card/site-card.tsx [hunk 1]
```diff
@@ -0,0 +1,45 @@
+import { Divider, IconButton } from '~/libs/components/components.js';
+
+import img from '../../../../assets/img/test.jpg';
```
Comment (dimapopovych) [inline frontend/src/pages/sites/components/site-card/site-card.tsx]: import

File: frontend/src/libs/components/components.ts [hunk 2]
```diff
@@ -14,4 +15,4 @@ export { Select } from './select/select.js';
 export { Stepper } from './stepper/stepper.js';
 export { Notification } from './toastr/notification.js';
 export { Provider as StoreProvider } from 'react-redux';
-export { Navigate, Outlet as RouterOutlet } from 'react-router-dom';
+export { Navigate, Outlet as RouterOutlet, useParams } from 'react-router-dom';
```
Comment (dimapopovych) [inline frontend/src/libs/components/components.ts]: useParams should be re-exported from hooks.ts

File: frontend/src/pages/sites/sites.tsx [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+import { useAppDispatch } from '~/libs/hooks/use-app-dispatch/use-app-dispatch.hook';
+import { type UserAuthResponse } from '~/packages/users/users.js';
+import { SiteCard } from '~/pages/sites/components/components.js';
+import { actions as sitesActions } from '~/slices/sites/sites.js';
+
+import styles from './styles.module.scss';
+
+const Sites: React.FC = () => {
+  const mockUser: UserAuthResponse = {
+    // replace with user from store
```
Comment (dimapopovych) [inline frontend/src/pages/sites/sites.tsx]:
```suggestion
    // replace with user from store
```

File: frontend/src/slices/sites/actions.ts [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+import { name as sliceName } from './sites.slice.js';
+
+const getSites = createAsyncThunk<
+  SiteGetAllResponseDto,
+  SiteGetByProjectRequestDtoType,
+  AsyncThunkConfig
+>(`${sliceName}/sites`, (payload, { extra }) => {
+  const { sitesApi } = extra;
+
+  return sitesApi.getProjectSites(payload.projectId);
```
Comment (dimapopovych) [inline frontend/src/slices/sites/actions.ts]: await?

File: frontend/src/slices/sites/sites.slice.ts [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+
+import { DataStatus } from '~/libs/enums/enums.js';
+import { type ValueOf } from '~/libs/types/types.js';
+import { type SiteGetAllResponseDto } from '~/packages/sites/sites.js';
+
+import { getSites } from './actions.js';
+
+type State = {
+  dataStatus: ValueOf<typeof DataStatus>;
+  sites: SiteGetAllResponseDto | null;
```
Comment (dimapopovych) [inline frontend/src/slices/sites/sites.slice.ts]:
```suggestion
  sites: SiteGetAllItemResponseDto[];
```

File: frontend/src/slices/sites/sites.slice.ts [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+};
+
+const initialState: State = {
+  dataStatus: DataStatus.IDLE,
+  sites: null,
```
Comment (dimapopovych) [inline frontend/src/slices/sites/sites.slice.ts]:
```suggestion
  sites: [];
```

File: backend/src/db/migrations/20230321180534_add_project_id_to_sites_table.ts [hunk 7]
```diff
@@ -0,0 +1,26 @@
+import { type Knex } from 'knex';
+
+const TABLE_NAME = 'sites';
+
+const ColumnName = {
+  PROJECT_ID: 'project_id',
+};
+
+function up(knex: Knex): Promise<void> {
+  return knex.schema.table(TABLE_NAME, (table) => {
```
Comment (andanf-e) [inline backend/src/db/migrations/20230321180534_add_project_id_to_sites_table.ts]:
```suggestion
  return knex.schema.alterTable(TABLE_NAME, (table) => {
```

File: shared/src/packages/sites/libs/validation-schemas/site-create.validation-schema.ts [hunk 8]
```diff
@@ -5,6 +5,7 @@ import { type SiteCreateRequestDto } from '../types/types.js';
 const siteCreate = joi.object<SiteCreateRequestDto, true>({
   name: joi.string().required(),
   publishedUrl: joi.string().allow(null),
+  projectId: joi.number().required(),
```
Comment (what1s1ove) [inline shared/src/packages/sites/libs/validation-schemas/site-create.validation-schema.ts]:
```suggestion
```
we should add it on the backend

File: shared/src/packages/sites/libs/validation-schemas/site-get-by-project.validation-schema.ts [hunk 9]
```diff
@@ -0,0 +1,9 @@
+import joi from 'joi';
+
+import { type SiteGetByProjectRequestDtoType } from '../types/types.js';
+
+const siteGetByProject = joi.object<SiteGetByProjectRequestDtoType, true>({
```
Comment (what1s1ove) [inline shared/src/packages/sites/libs/validation-schemas/site-get-by-project.validation-schema.ts]:
```suggestion
const siteGetByProjectParams = joi.object<SiteGetByProjectRequestDtoType, true>({
```
don't forget about file name

File: shared/src/packages/sites/libs/types/site-get-by-project-request-dto.type.ts [hunk 10]
```diff
@@ -0,0 +1,5 @@
+type SiteGetByProjectRequestDtoType = {
```
Comment (what1s1ove) [inline shared/src/packages/sites/libs/types/site-get-by-project-request-dto.type.ts]:
```suggestion
type SiteGetByProjectParamsDto = {
```
don't forget about file name

File: shared/src/packages/sites/libs/types/site-get-all-item-response-dto.type.ts [hunk 11]
```diff
@@ -2,6 +2,7 @@ type SiteGetAllItemResponseDto = {
   id: number;
   name: string;
   publishedUrl: string | null;
+  projectId: number;
```
Comment (what1s1ove) [inline shared/src/packages/sites/libs/types/site-get-all-item-response-dto.type.ts]:
```suggestion
```
should be added on the backend (not part of your task)

File: shared/src/packages/sites/libs/types/site-create-request-dto.type.ts [hunk 12]
```diff
@@ -1,6 +1,7 @@
 type SiteCreateRequestDto = {
   name: string;
   publishedUrl: string | null;
+  projectId: number;
```
Comment (what1s1ove) [inline shared/src/packages/sites/libs/types/site-create-request-dto.type.ts]: should be added on the backend (not part of your task)

File: shared/src/packages/sites/libs/enums/sites-api-path.enum.ts [hunk 13]
```diff
@@ -1,5 +1,6 @@
 const SitesApiPath = {
   ROOT: '/',
+  BY_PROJECT: '/:projectId',
```
Comment (what1s1ove) [inline shared/src/packages/sites/libs/enums/sites-api-path.enum.ts]:
```suggestion
  $PROJECT_ID: '/:projectId',
```
`$` for params

File: frontend/src/pages/sites/sites.tsx [hunk 14]
```diff
@@ -0,0 +1,51 @@
...
+
+import styles from './styles.module.scss';
+
+const mockUser: UserAuthResponse = {
+  id: 1,
+  email: 'pisotska.o@gmail.com',
+  firstName: 'Alyona',
+  lastName: 'Pisotska',
+  accountName: 'Pisotska',
+};
```
Comment (andanf-e) [inline frontend/src/pages/sites/sites.tsx]: Get user from store
```suggestion

```

File: frontend/src/slices/sites/actions.ts [hunk 15]
```diff
@@ -0,0 +1,21 @@
...
+
+import { type AsyncThunkConfig } from '~/libs/types/async-thunk-config.type.js';
+import {
+  type SiteGetAllResponseDto,
+  type SiteGetByProjectRequestDtoType,
+} from '~/packages/sites/sites.js';
+
+import { name as sliceName } from './sites.slice.js';
+
+const getSites = createAsyncThunk<
```
Comment (what1s1ove) [inline frontend/src/slices/sites/actions.ts]:
```suggestion
const getSitesByProject = createAsyncThunk<
```

**Review by andanf-e — CHANGES_REQUESTED** [review]

File: frontend/src/pages/sites/sites.tsx [hunk 16]
```diff
@@ -0,0 +1,51 @@
...
+    }
+  }, [dispatch, projectId]);
+
+  const { sites } = useAppSelector((state) => ({
+    sites: state.sites.sites,
+  }));
+
+  if (projectId === undefined) {
+    return <Navigate to={AppRoute.PROJECTS} replace />;
+  }
```
Comment (what1s1ove) [inline frontend/src/pages/sites/sites.tsx]:
```suggestion
```
don't add this check for now

File: frontend/src/pages/my-projects/components/project-card/project-card.tsx [hunk 17]
```diff
@@ -1,16 +1,24 @@
...
 
-const ProjectCard: React.FC<Properties> = ({ siteName }: Properties) => {
+const ProjectCard: React.FC<Properties> = ({ siteName, id }: Properties) => {
   return (
     <div className={styles['card']}>
       <img className={styles['card-image']} src="imgLink" alt="" />
       <div className={styles['card-description']}>
-        <h3>{siteName}</h3>
+        <Link
+          to={`${ProjectsDetailsRoute.root}/${id}${ProjectsDetailsRoute.sites}`}
```
Comment (what1s1ove) [inline frontend/src/pages/my-projects/components/project-card/project-card.tsx]: use `configurateString` helper for it

File: frontend/src/packages/sites/sites-api.ts [hunk 18]
```diff
@@ -0,0 +1,35 @@
...
+  http: IHttp;
+  storage: IStorage;
+};
+
+class SitesApi extends HttpApi {
+  public constructor({ baseUrl, http, storage }: Constructor) {
+    super({ path: ApiPath.SITES, baseUrl, http, storage });
+  }
+
+  public async getProjectSites(
```
Comment (what1s1ove) [inline frontend/src/packages/sites/sites-api.ts]:
```suggestion
  public async getByProectId(
```

File: frontend/src/libs/enums/app-route.enum.ts [hunk 19]
```diff
@@ -1,10 +1,17 @@
+const ProjectsDetailsRoute = {
+  root: '/projects',
+  $id: '/:projectId',
+  sites: '/sites',
+};
```
Comment (what1s1ove) [inline frontend/src/libs/enums/app-route.enum.ts]:
```suggestion
```
descibe all in AppRoute

File: frontend/src/libs/enums/app-route.enum.ts [hunk 20]
```diff
@@ -1,10 +1,17 @@
...
+
 const AppRoute = {
   ROOT: '/',
   SIGN_IN: '/sign-in',
   SIGN_UP: '/sign-up',
-  PROJECTS: '/projects',
+  PROJECTS: ProjectsDetailsRoute.root,
+  SITES: `${ProjectsDetailsRoute.root}${ProjectsDetailsRoute.$id}${ProjectsDetailsRoute.sites}`,
```
Comment (what1s1ove) [inline frontend/src/libs/enums/app-route.enum.ts]:
```suggestion
  PROJECTS: '/projects',
  PROJECTS_$PROJECT_ID_SITES: '/projects/:projectId/sites'
```

File: frontend/src/assets/img/test.jpg
Comment (what1s1ove) [inline frontend/src/assets/img/test.jpg:1]: site-mock-image

File: backend/src/packages/sites/site.service.ts [hunk 21]
```diff
@@ -23,6 +23,16 @@ class SiteService implements Omit<IService, 'find' | 'update' | 'delete'> {
     };
   }
 
+  public async findByProject(
```
Comment (what1s1ove) [inline backend/src/packages/sites/site.service.ts]:
```suggestion
  public async findByProjectId(
```

File: backend/src/packages/sites/site.repository.ts [hunk 22]
```diff
@@ -17,12 +17,21 @@ class SiteRepository
     return sites.map((site) => SiteEntity.initialize(site));
   }
 
+  public async findByProject(projectId: number): Promise<SiteEntity[]> {
```
Comment (what1s1ove) [inline backend/src/packages/sites/site.repository.ts]:
```suggestion
  public async findByProjectId(projectId: number): Promise<SiteEntity[]> {
```

File: backend/src/packages/sites/site.controller.ts [hunk 23]
```diff
@@ -84,6 +104,34 @@ class SiteController extends Controller {
...
+   *             schema:
+   *               type: object
+   *               properties:
+   *                 items:
+   *                   type: array
+   *                   items:
+   *                     $ref: '#/components/schemas/Site'
+   *                   minItems: 0
+   */
+  private async findByProject(
```
Comment (what1s1ove) [inline backend/src/packages/sites/site.controller.ts]:
```suggestion
  private async findByProjectId(
```

**Review by what1s1ove — CHANGES_REQUESTED** [review]

File: frontend/src/assets/css/variables.scss [hunk 24]
```diff
@@ -12,5 +12,8 @@
   --blue: #3b79c3;
   --text-type-400: #130138;
   --text-type-300: #5a6474;
+  --card-widh: 294px;
+  --card-height: 296px;
+  --card-gap: 20px;
```
Comment (andanf-e) [inline frontend/src/assets/css/variables.scss]: That's nice to have reused things be moved to commons, but I believe `variables` should contain only global dependencies for styles -- so let's have this values declared locally for needed styles
```suggestion

```

File: frontend/src/pages/my-projects/components/project-card/project-card.tsx (hunk 17 above)
Comment (Alyona-Pisotska) [reply frontend/src/pages/my-projects/components/project-card/project-card.tsx]: @what1s1ove  I did it. But I have Type problem with the Link component. Property "to" of Link component has AppRoute type:
`type Properties = {
  to: ValueOf<typeof AppRoute>;
  children: React.ReactNode;
};`

But configureString returns just string type. So maybe I need to use the NavLink component or create a new Link component that has the property "to" with string type or something else ...

File: .github/ISSUE_TEMPLATE/bug-report.md
Comment (andanf-e) [inline .github/ISSUE_TEMPLATE/bug-report.md:1]: should not be removed

File: .ls-lint.yml [hunk 25]
```diff
@@ -14,8 +14,6 @@ ls:
   .woff2: regex:([.a-z0-9]*)([-.][a-z0-9]+)*
   backend/src/db/migrations:
     .ts: snake_case
-  .github/ISSUE_TEMPLATE:
```
Comment (andanf-e) [inline .ls-lint.yml]: turn it back

File: backend/src/packages/sites/site.repository.ts [hunk 26]
```diff
@@ -17,6 +17,15 @@ class SiteRepository
     return sites.map((site) => SiteEntity.initialize(site));
   }
 
+  public async findByProjectId(projectId: number): Promise<SiteEntity[]> {
+    const sites = await this.siteModel
+      .query()
+      .where('project_id', projectId)
```
Comment (andanf-e) [inline backend/src/packages/sites/site.repository.ts]: Within the snakeCaseMapper we may define keys in non-snake_case(snakeCase) for not a raw queries

File: frontend/src/pages/sites/sites.tsx [hunk 27]
```diff
@@ -0,0 +1,43 @@
...
+    return auth.user as UserAuthResponse;
+  });
+
+  const { sites } = useAppSelector((state) => ({
+    sites: state.sites.sites,
+  }));
+
+  return (
+    <>
+      <Header user={user} pageName={PageName.MY_SITES} />
```
Comment (andanf-e) [inline frontend/src/pages/sites/sites.tsx]: Not sure if we should have header in pages other than root page

File: frontend/src/pages/sites/sites.tsx [hunk 28]
```diff
@@ -0,0 +1,41 @@
...
+
+import styles from './styles.module.scss';
+
+const Sites: React.FC = () => {
+  const { projectId } = useParams();
+  const dispatch = useAppDispatch();
+
+  useEffect((): void => {
+    if (projectId) {
+      void dispatch(sitesActions.getSitesByProject({ projectId: +projectId }));
```
Comment (andanf-e) [inline frontend/src/pages/sites/sites.tsx]:
```suggestion
      void dispatch(sitesActions.getSitesByProject({ projectId: Number(projectId) }));
```

File: frontend/src/pages/sites/sites.tsx [hunk 29]
```diff
@@ -0,0 +1,41 @@
...
+    }
+  }, [dispatch, projectId]);
+
+  const { sites } = useAppSelector((state) => ({
+    sites: state.sites.sites,
+  }));
```
Comment (andanf-e) [inline frontend/src/pages/sites/sites.tsx]:
```suggestion
  const { sites } = useAppSelector(({ sites }) => ({
    sites: sites.sites,
  }));
```

File: frontend/src/pages/my-projects/components/project-card/project-card.tsx [hunk 30]
```diff
@@ -1,17 +1,30 @@
...
+import { configureString } from '~/libs/helpers/helpers.js';
+import { type ValueOf } from '~/libs/types/types.js';
+
 import styles from './styles.module.scss';
 
 type Properties = {
-  id?: number;
   siteName: string;
   userId?: number;
+  id: number;
```
Comment (what1s1ove) [inline frontend/src/pages/my-projects/components/project-card/project-card.tsx]:
```suggestion
  project: Type
```
pass a full project to props

File: frontend/src/pages/sites/components/site-card/site-card.tsx [hunk 31]
```diff
@@ -0,0 +1,46 @@
+import img from 'src/assets/img/site-mock-image.jpg';
+
+import { Divider, IconButton } from '~/libs/components/components.js';
+
+import styles from './styles.module.scss';
+
+type Properties = {
+  key: number;
+  siteName: string;
```
Comment (what1s1ove) [inline frontend/src/pages/sites/components/site-card/site-card.tsx]:
```suggestion
  site: Type
```
pass a full project to props

File: shared/src/packages/sites/libs/enums/sites-api-path.enum.ts [hunk 32]
```diff
@@ -1,5 +1,6 @@
 const SitesApiPath = {
   ROOT: '/',
+  $PROJECT_ID: '/:projectId',
```
Comment (what1s1ove) [inline shared/src/packages/sites/libs/enums/sites-api-path.enum.ts]:
```suggestion
```
the route should be `/projects/:projectId/sites` and it's a part of the `ProjectsApiPath `

**Review by what1s1ove — CHANGES_REQUESTED** [review]

**Review by what1s1ove — APPROVED** [review]

**Review by andanf-e — APPROVED** [review]

**Pull request merged**